### PR TITLE
net: pkt: context: Respect reserved bytes

### PR DESCRIPTION
**Pull request by rettichschnidi**
When operating on a net_pkt using the curser functionality, reserved
bytes (headroom) must be respected.

This commit achieves this by updating the size member of the
net_buf_simple structure when reserving (and pushing) bytes.

While this solution changes the semantic of the size member, it requires
a low amount of code changes in addition to allowing the pattern of
writing (up to) size bytes starting at the data pointer to work properly
even in the presence of reserved bytes.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

### CODE REVIEW

File: include/net/buf.h [hunk 1]
```diff
@@ -91,7 +91,7 @@ struct net_buf_simple {
 	/** Length of the data behind the data pointer. */
```
Comment (jukkar) [inline include/net/buf.h:91]: typo in the commit msg, s/curser/cursor/
Also you can shorten the commit subject as the "context" is not really needed here, just write it
`net: pkt: Respect reserved bytes`

**Review by jukkar — COMMENTED** [review]
This looks reasonable.
@jhedberg does this cause issues in Bluetooth side?

File: include/net/buf.h (hunk 1 above)
Comment (rettichschnidi) [reply include/net/buf.h:91]: fixed

**Comment by jhedberg** [comment]
I need to look more carefully at this. Meanwhile, please split the net_buf changes into a separate commit with a `net: buf: ...` prefix. The net_buf semantics and API terminology has been created by using Linux sk_buff as reference - I think we should keep the same semantics in order not to confuse people. The net_buf API is not a 1:1 mapping to sk_buff, so I'll need to check if this is true for the parts that this PR affects.

**Comment by rettichschnidi** [comment]
Just reworked the PR.

I guess all of this boils down to two questions:
* Is it desirable to use the net_pkt cursor as done in the first commit? (I think so)
* Is the solution proposed in the 2nd commit the right/best one? (Unsure about this)

**Comment by jhedberg** [comment]
>  In order to (safely) append data to a net_buf object, the user must query net_buf_headroom() to ensure enough space is left.

@rettichschnidi I don't understand this comment. To check if there's space to append data you only need to query `net_buf_tailroom()`

**Comment by jhedberg** [comment]
@rettichschnidi I get the feeling that the background of this PR is the fact that you've been using `buf->size` as a primary API, when it was mainly designed to be useful only for net_buf internal purposes. From the net_buf API user perspective the `net_buf_headroom()` and `net_buf_tailroom()` are the primary methods that are expected to be used in order to determine if there's space to append or prepend more data to the buffer.

**Comment by rettichschnidi** [comment]
> @rettichschnidi I don't understand this comment. To check if there's space to append data you only need to query `net_buf_tailroom()`

My comment is misleading. I will clarify this with an updated comment.

> @rettichschnidi I get the feeling that the background of this PR is the fact that you've been using `buf->size` as a primary API, when it was mainly designed to be useful only for net_buf internal purposes. From the net_buf API user perspective the `net_buf_headroom()` and `net_buf_tailroom()` are the primary methods that are expected to be used in order to determine if there's space to append or prepend more data to the buffer.

Yes, the problem stems from the (ab)use of `buf->size`. However, it is not (just) me, but also existing code in the Zephyr code base, like the code implementing the net_pkt_cursor manipulations.

I will come up with an updated proposal which does not not change the (the semantic of the) size member of net_buf, but using it properly instead.

**Review by jukkar — APPROVED** [review]

**Comment by rettichschnidi** [comment]
I am unsure why buildkite fails - neither `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T tests/kernel/lifo/lifo_usage/` nor `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T samples/kernel/condition_variables/condvar/` errors on my own machine.

**Comment by jukkar** [comment]
> I am unsure why buildkite fails - neither `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T tests/kernel/lifo/lifo_usage/` nor `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T samples/kernel/condition_variables/condvar/` errors on my own machine.

There seems to be some unrelated issues in the kernel side and some tests are indeed failing. Dunno what is wrong there.
But there is one net_pkt failure here
```
/workdir/zephyr/subsys/net/ip/net_pkt.c: In function 'net_pkt_print_frags':
/workdir/zephyr/subsys/net/ip/net_pkt.c:347:45: error: 'buf' undeclared (first use in this function)
  347 |     count, frag, frag->len, net_buf_max_len(buf),
      |                                             ^~~
```

**Review by jhedberg — APPROVED** [review]

**Pull request merged**